### PR TITLE
Improve the user experience in Azure

### PR DESCRIPTION
**Pull request by hui1110**
Include the following:

* `Azure` specific executor extensions, add implementation of new scenarios.

* Add `Azure` related guides to `HELP.md` to help users get more information.

### CODE REVIEW

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+	public SpringAzureHelpDocumentCustomizer(Build build) {
+		this.build = build;
+	}
+
+	@Override
+	public void customize(HelpDocument document) {
+		if (this.build.dependencies().has(AZURE_DEPENDENCY_ID)) {
+			document.gettingStarted().addGuideLink(
+					"https://docs.microsoft.com/en-us/azure/developer/java/spring-framework/spring-boot-starters-for-azure",
```
Comment (backwind1233) [inline start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:56]: remove en-us

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+ * A {@link HelpDocumentCustomizer} that Getting Started Guides.
+ *
+ * @author Yonghui Ye
+ */
+public class SpringAzureHelpDocumentCustomizer implements HelpDocumentCustomizer {
+
+	private static final String AZURE_DEPENDENCY_ID = "azure-support";
+
```
Comment (backwind1233) [inline start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:31]: It's better to remove blank lines here

File: start-site/src/main/resources/application.yml [hunk 3]
```diff
@@ -1479,10 +1479,10 @@ initializr:
               href: https://aka.ms/spring/samples/latest/cosmos
               description: Azure Cosmos DB Sample
         - name: Azure Key Vault
-          id: azure-keyvault-secrets
+          id: azure-keyvault
           groupId: com.azure.spring
-          artifactId: spring-cloud-azure-starter-keyvault-secrets
-          description: Manage application secrets.
+          artifactId: spring-cloud-azure-starter-keyvault
```
Comment (backwind1233) [inline start-site/src/main/resources/application.yml:1484]: Do we have "spring-cloud-azure-starter-keyvault"?

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureDefaultBuildCustomizer.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * @author Yonghui Ye
+ */
+public class SpringAzureDefaultBuildCustomizer implements BuildCustomizer<Build> {
+
+	@Override
+	public void customize(Build build) {
+		if (build.dependencies().items().anyMatch((u) -> u.getGroupId().equals("com.azure.spring")) && !build
+				.dependencies().items().anyMatch((u) -> u.getArtifactId().equals("spring-cloud-azure-starter"))) {
+			build.dependencies().add("azure-support");
```
Comment (wilkinsona) [inline start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureDefaultBuildCustomizer.java:34]: Why is this necessary? Every `spring-cloud-azure-starter-*` module should depend on `spring-cloud-azure-starter` just as each of Spring Boot's own starters depends on `spring-boot-starter`.

**Review by wilkinsona — CHANGES_REQUESTED** [review]

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureSleuthBuildCustomizer.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package io.spring.start.site.extension.dependency.springazure;
+
+import io.spring.initializr.generator.buildsystem.Build;
+import io.spring.initializr.generator.buildsystem.Dependency;
+import io.spring.initializr.generator.spring.build.BuildCustomizer;
+
+/**
+ * A {@link BuildCustomizer}that automatically adds "spring-cloud-azure-trace-sleuth" when
```
[the remark this answers is not in the document]
Comment (hui1110) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureSleuthBuildCustomizer.java]: added

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/package-info.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Extensions for generation of projects that depend on Spring Batch.
```
[the remark this answers is not in the document]
Comment (hui1110) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/package-info.java]: updated

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureActuatorBuildCustomizer.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import io.spring.initializr.generator.buildsystem.Dependency;
+import io.spring.initializr.generator.spring.build.BuildCustomizer;
+
+/**
+ * A {@link BuildCustomizer} that automatically adds "spring-cloud-azure-starter-actuator"
+ * when any Spring Cloud Azure library and Spring Boot Actuator are selected.
+ *
+ * @author Yonghui Ye
+ */
+public class SpringAzureActuatorBuildCustomizer implements BuildCustomizer<Build> {
```
[the remark this answers is not in the document]
Comment (hui1110) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureActuatorBuildCustomizer.java]: updated

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureDefaultBuildCustomizer.java (hunk 4 above)
Comment (wilkinsona) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureDefaultBuildCustomizer.java:34]: Sorry, that's not quite what I meant. This whole customiser should not be necessary as each `spring-cloud-azure-starter-*` should declare a dependency on `spring-cloud-azure-starter` in its pom.

How would we get into a situation where there's a `com.azure.spring:*` dependency but no `spring-cloud-azure-starter-*` dependency?

**Review by wilkinsona — CHANGES_REQUESTED** [review]
Thanks for the updates. I'm still not sure about `SpringAzureDefaultBuildCustomizer`.

**Comment by hui1110** [comment]
> Thanks for the updates. I'm still not sure about `SpringAzureDefaultBuildCustomizer`.

> Thanks for your feedback. There really is no such scenario at the moment, `SpringAzureDefaultBuildCustomizer` is unnecessary.

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureDefaultBuildCustomizer.java (hunk 4 above)
Comment (hui1110) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureDefaultBuildCustomizer.java]: Thanks for your feedback. There really is no such scenario at the moment, `SpringAzureDefaultBuildCustomizer` is unnecessary.

File: start-site/src/main/resources/application.yml [hunk 8]
```diff
@@ -109,7 +109,7 @@ initializr:
         versionProperty: spring-cloud-azure.version
         mappings:
           - compatibilityRange: "[2.5.0.M1,3.0.0-M1)"
-            version: 4.3.0
+            version: 4.2.0
```
Comment (wilkinsona) [inline start-site/src/main/resources/application.yml:112]: Why are we no longer upgrading to 4.3.0?

Comment (hui1110) [reply start-site/src/main/resources/application.yml:112]: I created this [issue ](https://github.com/spring-io/start.spring.io/issues/955)to upgrade the bom version. Maybe it can be done with this pr？

Comment (wilkinsona) [reply start-site/src/main/resources/application.yml:112]: Given the title and description of this PR, I had expected it to include the upgrade to 4.3.0. It's completely fine if you'd prefer to separate the two, but the title and description of this PR should be updated. It looks it is now focussed solely on improving the Azure experience so you get better HELP.md contents, Azure-specific actuator extensions, etc.

Comment (hui1110) [reply start-site/src/main/resources/application.yml:112]: Updated, please take a look.

**Comment by hui1110** [comment]
Hi @wilkinsona, do you have any other questions about this PR? Can it be merged?

**Comment by wilkinsona** [comment]
Thanks for your patience. I don't have any further questions at the moment but haven't had the time to review the latest change in detail. I will do so as soon as priorities allow.

**Comment by wilkinsona** [comment]
I should have some time on Friday or next week. This week's Spring Boot releases are taking priority at the moment.

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import io.spring.initializr.generator.buildsystem.Build;
+import io.spring.initializr.generator.spring.documentation.HelpDocument;
+import io.spring.initializr.generator.spring.documentation.HelpDocumentCustomizer;
+
+/**
+ * A {@link HelpDocumentCustomizer} that Getting Started Guides.
+ *
+ * @author Yonghui Ye
+ */
+class SpringAzureHelpDocumentCustomizer implements HelpDocumentCustomizer {
```
Comment (wilkinsona) [inline start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:28]: I'm not sure that this class is needed. The links could be declared in `application.yml` instead using `rel: reference` or `rel: guide` as appropriate. I think the link descriptions should reflect the titles of the pages to which they link.

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+	SpringAzureHelpDocumentCustomizer(Build build) {
+		this.build = build;
+	}
+
+	@Override
+	public void customize(HelpDocument document) {
+		addMsDocsLink(document, AZURE_DEPENDENCY_ID, "https://aka.ms/spring/msdocs", "Spring Cloud Azure");
+		addMsDocsLink(document, AZURE_ACTIVE_DIRECTORY_DEPENDENCY_ID, "https://aka.ms/spring/msdocs/aad",
+				"Azure Active Directory");
```
Comment (wilkinsona) [inline start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:56]: As an example of what I mean about the link descriptions, I think this one should be "Secure a Java web app using the Spring Boot Starter for Azure Active Directory"

**Review by wilkinsona — CHANGES_REQUESTED** [review]

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java (hunk 10 above)
Comment (hui1110) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:56]: I will update them.

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+	private final Build build;
+
+	SpringAzureHelpDocumentCustomizer(Build build) {
+		this.build = build;
+	}
+
+	@Override
+	public void customize(HelpDocument document) {
+		addMsDocsLink(document, AZURE_DEPENDENCY_ID, "https://aka.ms/spring/msdocs", "Spring Cloud Azure");
```
[the remark this answers is not in the document]
Comment (hui1110) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:54]: It does look like the reference documentation. They are in different places and show different content.

File: start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java (hunk 9 above)
Comment (hui1110) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:28]: I will update them.

Comment (hui1110) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:28]: When `spring-cloud-azure-starter-actuator`, `spring-cloud-azure-trace-sleuth`, `spring-cloud-azure-starter-integration-storage-queue` appear in the `pom`, the corresponding document cannot use `rel: reference` or `rel : guide` to add. This class should be needed.

Comment (wilkinsona) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:28]: Thanks. I agree – for those dependencies that are added implicitly due to other dependencies having been selected we do need to do something. It looks like implicit dependencies ([as we now use for Testcontainers](https://github.com/spring-io/start.spring.io/commit/1cfafcf0b3ab761bcaddbffb779ab1f65867031d)) would be a good fit.

Comment (wilkinsona) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:28]: Do think it makes sense to use implicit dependencies, @snicoll?

Comment (snicoll) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:28]: Looks like it, yes!

Comment (hui1110) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/springazure/SpringAzureHelpDocumentCustomizer.java:28]: @wilkinsona. Does `SpringAzureHelpDocumentCustomizer` need to be changed? We just need to add one or more links in `HELP.md`.

**Comment by hui1110** [comment]
Hi @wilkinsona, can you help take another look?

**Comment by hui1110** [comment]
> I have polished the proposal to use implicit dependencies. You can see the changes in [this branch](https://github.com/wilkinsona/start.spring.io/tree/gh-946).
> 
> There's one thing left that I don't think is quite right. At the moment, an app that's using Azure links to both https://microsoft.github.io/spring-cloud-azure/current/reference/html/index.html and https://aka.ms/spring/msdocs. Do we really need both? The two pages are titled "Spring Cloud Azure - Reference Documentation" and "Spring Cloud Azure documentation" so they appear to be very similar.

Thank you very much for your proposal. The titles of the two documents do look very similar. https://microsoft.github.io/spring-cloud-azure/current/reference/html/index.html is our community edition documentation and https://aka.ms/spring/msdocs is our official documentation. We would like to provide both documents in `HELP.md`. If you don't think this is right, we'll consider removing one of them.

**Comment by wilkinsona** [comment]
At the moment, we've got these entries in the yaml file:

```yaml
- rel: reference
  href: https://microsoft.github.io/spring-cloud-azure/current/reference/html/index.html
  description: Spring Cloud Azure Reference Documentation
- rel: reference
  href: https://aka.ms/spring/msdocs
  description: Spring Cloud Azure documentation - Microsoft Docs
```

I find the two confusing as I can't tell which I should read to learn certain things. It leaves me wondering what the difference is between "documentation" and "reference documentation" and why, given that Azure is a Microsoft offering, is there Microsoft documentation and "plain" documentation. I'm not opposed to having multiple links, but I think it's important that they help a user by guiding them to something useful rather than confusing them.

**Comment by hui1110** [comment]
> At the moment, we've got these entries in the yaml file:
> 
> ```yaml
> - rel: reference
>   href: https://microsoft.github.io/spring-cloud-azure/current/reference/html/index.html
>   description: Spring Cloud Azure Reference Documentation
> - rel: reference
>   href: https://aka.ms/spring/msdocs
>   description: Spring Cloud Azure documentation - Microsoft Docs
> ```
> 
> I find the two confusing as I can't tell which I should read to learn certain things. It leaves me wondering what the difference is between "documentation" and "reference documentation" and why, given that Azure is a Microsoft offering, is there Microsoft documentation and "plain" documentation. I'm not opposed to having multiple links, but I think it's important that they help a user by guiding them to something useful rather than confusing them.

Updated.

**Comment by hui1110** [comment]
Hi @wilkinsona, please take a look when you are free.

**Comment by wilkinsona** [comment]
Thanks for the updates, @hui1110. The changes have now been merged.
